### PR TITLE
core improvements for clustering module

### PR DESCRIPTION
**Pull request by groldan**

### CODE REVIEW

**Comment by jodygarnett** [comment]
@groldan does this replace #860 (which was against 2.6.x)?

**Comment by smithkm** [comment]
@jodygarnett I think this is the promised pull to master that #860 was a preemptive backport of.

File: src/main/src/main/java/org/geoserver/catalog/ResourcePool.java [hunk 1]
```diff
@@ -1944,8 +1959,12 @@ protected void dispose(String key, FeatureType featureType) {
             String id = key.substring(0, key.indexOf(PROJECTION_POLICY_SEPARATOR));
         	FeatureTypeInfo info = catalog.getFeatureType(id);
             if(info != null){
-                LOGGER.info( "Disposing feature type '" + info.getName() + "'");
+                LOGGER.info( "Disposing feature type '" + info.getName() + "'/" + id);
                 fireDisposed(info, featureType);
+                if (null != featureTypeAttributeCache.remove(id)) {
+                    LOGGER.info("AttributeType cache cleared for feature type '" + info.getName()
```
Comment (jodygarnett) [inline src/main/src/main/java/org/geoserver/catalog/ResourcePool.java]: suspect I can knock this down to FINE visibility.

File: src/wfs/src/main/java/org/geoserver/wfs/CatalogFeatureTypeCache.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
+/* Copyright (c) 2014 OpenPlans - www.openplans.org. All rights reserved.
```
Comment (jodygarnett) [inline src/wfs/src/main/java/org/geoserver/wfs/CatalogFeatureTypeCache.java]: Need to update to OSGeo header

File: src/wfs/src/main/java/org/geoserver/wfs/xml/v1_0_0/WFSConfiguration.java [hunk 3]
```diff
@@ -168,28 +166,7 @@ public void configureContext(MutablePicoContainer context) {
         
         //TODO: this code is copied from the 1.1 configuration, FACTOR IT OUT!!!
         //seed the cache with entries from the catalog
-        FeatureTypeCache featureTypeCache = (FeatureTypeCache) context
```
Comment (smithkm) [inline src/wfs/src/main/java/org/geoserver/wfs/xml/v1_0_0/WFSConfiguration.java]: Does this change make this comment irrelevant?

File: src/wfs/src/main/java/org/geoserver/wfs/xml/FeatureTypeSchemaBuilder.java [hunk 4]
```diff
@@ -192,13 +194,27 @@ public final XSDSchema buildSchemaInternal(FeatureTypeInfo[] featureTypeInfos, S
...
+                    simple = featureTypeInfos[i].getFeatureType() instanceof SimpleFeatureType;
+                }
+                catch(IOException e) {
+                    // ignore so that broken feature types don't prevent others from continuing to work
                 }
             }
+
+            if (!simple) {
+                // add secondary namespaces from catalog
+                for (NamespaceInfo nameSpaceinfo : catalog.getNamespaces()) {
```
Comment (smithkm) [inline src/wfs/src/main/java/org/geoserver/wfs/xml/FeatureTypeSchemaBuilder.java]: Full catalog scan of namespaces.  It was here before so it's not making it worse, but this is something to look at for future improvement.

**Comment by jodygarnett** [comment]
Accepting this one with minor modifications, @smithkm has also reviewed.

**Pull request merged**